pkg/project/auth/watch.go

```diff
@@ -0,0 +1,217 @@
+)
+
+type CacheWatcher interface {
+	// GroupMembershipChanged is called serialially for all changes for all watchers.  This method MUST NOT BLOCK.
+	// The serial nature makes reasoning about the code easy, but if you block in this method you will doom all watchers.
+	GroupMembershipChanged(namespaceName string, latestUsers, latestGroups, removedUsers, removedGroups, addedUsers, addedGroups sets.String)
+}
+
+type WatchableCache interface {
+	// RemoveWatcher removes a watcher
+	RemoveWatcher(CacheWatcher)
+	// List returns the set of namespace names the user has access to view
+	List(userInfo user.Info) (*kapi.NamespaceList, error)
+}
+
+// userProjectWatcher converts a native etcd watch to a watch.Interface.
+type userProjectWatcher struct {
+	username string
+	groups   []string
+
+	// cacheIncoming is a buffered channel used for notification to watcher.  If the buffer fills up,
+	// then the watcher will be removed and the connection will be broken.
+	cacheIncoming chan watch.Event
+	// cacheError is a cached channel that is put to serialially.  In theory, only one item will
+	// ever be placed on it.
+	cacheError chan error
+
+	// outgoing is the unbuffered `ResultChan` use for the watch.  Backups of this channel will block
+	// the default `emit` call.  That's why cacheError is a buffered channel.
+	outgoing chan watch.Event
+	// userStop lets a user stop his watch.
+	userStop chan struct{}
+
+	// stopped prevents closing closed channels
+	stopped bool
+	// stopLock keeps parallel stops from doing crazy things
+	stopLock sync.Mutex
+
+	// Injectable for testing. Send the event down the outgoing channel.
+	emit func(watch.Event)
+
+	projectCache *projectcache.ProjectCache
+	authCache    WatchableCache
+
+	knownProjects sets.String
+}
+
+var (
+	watchChannelHWM etcd.HighWaterMark
+)
+
+func NewUserProjectWatcher(username string, groups []string, projectCache *projectcache.ProjectCache, authCache WatchableCache) *userProjectWatcher {
+	userInfo := &user.DefaultInfo{Name: username, Groups: groups}
+	namespaces, _ := authCache.List(userInfo)
+	knownProjects := sets.String{}
+	for _, namespace := range namespaces.Items {
+		knownProjects.Insert(namespace.Name)
+	}
+
+	w := &userProjectWatcher{
+		username: username,
+		groups:   groups,
+
+		cacheIncoming: make(chan watch.Event, 1000),
+		cacheError:    make(chan error, 1),
+		outgoing:      make(chan watch.Event),
+		userStop:      make(chan struct{}),
+
+		projectCache:  projectCache,
+		authCache:     authCache,
+		knownProjects: knownProjects,
+	}
+	w.emit = func(e watch.Event) {
+		w.outgoing <- e
+	}
+	return w
+}
+
+func (w *userProjectWatcher) GroupMembershipChanged(namespaceName string, latestUsers, lastestGroups, removedUsers, removedGroups, addedUsers, addedGroups sets.String) {
+	removed := removedUsers.Has(w.username) || removedGroups.HasAny(w.groups...)
+	added := addedUsers.Has(w.username) || addedGroups.HasAny(w.groups...)
```

Comment:
do you even need to track `added` (I'm not seeing it used apart from `hasAccess`)?